### PR TITLE
[ONNX] Added support and test cases for Shape-15 

### PR DESCRIPTION
**Pull request by MeeCreeps**
related issue : #20194

### CODE REVIEW

File: src/frontends/onnx/tests/onnx_import.in.cpp [hunk 1]
```diff
@@ -1800,6 +1800,108 @@ OPENVINO_TEST(${BACKEND_NAME}, onnx_model_shape) {
...
+
+    auto test_case = ov::test::TestCase(model, s_device);
+    test_case.add_multiple_inputs(inputs);
+    test_case.add_expected_output<int64_t>({3});
+    test_case.run();
+}
+
+OPENVINO_TEST(${BACKEND_NAME}, onnx_model_shape15_end_negative_1) {
+    // shape 15 tests
+    auto model = convert_model("shape15.onnx");
```
Comment (gkrivor) [inline src/frontends/onnx/tests/onnx_import.in.cpp:1822]: I have a strong feeling it is copy-paste issue :) Could you check filenames? I see you introduced several test models, but don't use them.

**Review by gkrivor — CHANGES_REQUESTED** [review]
Hi @MeeCreeps , thank you for contribution!

Please, check important comment.

And also, could you take a look, I may expect these test cases should work after your change:
https://github.com/openvinotoolkit/openvino/blob/c208a881cb4e1889d8d7126b838b4fcb86b44b8d/src/frontends/onnx/tests/tests_python/test_backend.py#L337

File: src/frontends/onnx/tests/onnx_import.in.cpp (hunk 1 above)
Comment (MeeCreeps) [reply src/frontends/onnx/tests/onnx_import.in.cpp]: sorry, I have corrected it

File: src/frontends/onnx/frontend/src/op/shape.cpp [hunk 2]
```diff
@@ -1,17 +1,49 @@
-// Copyright (C) 2018-2024 Intel Corporation
+// Copyright (C) 2018-2023 Intel Corporation
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/shape.cpp:1]: Please, do not change copyrights this way :)

**Comment by mlukasze** [comment]
please fix clang to pass a test

File: src/frontends/onnx/frontend/src/op/shape.cpp [hunk 3]
```diff
@@ -1,17 +1,49 @@
...
+    auto end_val = node.get_attribute_value<int64_t>("end", -1);
+
+    start_val = start_val < 0 ? -1 : start_val;
+    end_val = end_val < 0 ? -1 : end_val;
+
+    auto start = v0::Constant::create(ov::element::i64, ov::Shape{1}, {start_val});
+    auto end = v0::Constant::create(ov::element::i64, ov::Shape{1}, {end_val});
+    const auto& default_step = v0::Constant::create(ov::element::i32, {1}, {1});
+    ov::Output<ov::Node> steps =
+        std::make_shared<v3::Broadcast>(default_step, std::make_shared<v3::ShapeOf>(start, ov::element::i64));
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/shape.cpp:40]: ShapeOf(start) looks strange for me, because it is known and const. Could you check this place?

**Comment by gkrivor** [comment]
@MeeCreeps please, take a look on test problems. Looks like there is some problems...

**Comment by MeeCreeps** [comment]
> @MeeCreeps please, take a look on test problems. Looks like there is some problems...

ok, i will check it out

**Comment by MeeCreeps** [comment]
@gkrivor , hello, i solved the problem and all tests are passed locally, can you check it out ? thanks

**Comment by mlukasze** [comment]
hey @MeeCreeps 
what happened?

**Comment by MeeCreeps** [comment]
> hey @MeeCreeps what happened?

sorry, I revert the branch and recommit changes, it's ready to review now.

**Review by mitruska — CHANGES_REQUESTED** [review]
Hello @MeeCreeps, thank you for this contribution!

Usage of Slice to resolve the issue is correct, the solution is almost ready, just few things to be fixed. I've also added optimization proposals.

File: src/frontends/onnx/frontend/src/op/shape.cpp (hunk 3 above)
Comment (mitruska) [reply src/frontends/onnx/frontend/src/op/shape.cpp]: Agree, `default_step` defined above should be sufficient, the `steps` as `Broadcast` + `ShapeOf` are rather redundant.

File: src/frontends/onnx/frontend/src/op/shape.cpp [hunk 4]
```diff
@@ -4,14 +4,44 @@
...
+namespace set_15 {
+
+ov::OutputVector shape(const ov::frontend::onnx::Node& node) {
+    using ov::op::util::is_null;
+    ov::OutputVector inputs{node.get_ov_inputs()};
+    const auto& data = inputs.at(0);
+    auto input_shape = std::make_shared<v3::ShapeOf>(data);
+
+    auto start_val = node.get_attribute_value<int64_t>("start", 0);
+    auto end_val = node.get_attribute_value<int64_t>("end", INT64_MAX);
```
[the remark this answers is not in the document]
Comment (MeeCreeps) [reply src/frontends/onnx/frontend/src/op/shape.cpp:31]: Thanks for your advice. In my new push, I fixed it all.

File: src/frontends/onnx/frontend/src/op/shape.cpp [hunk 5]
```diff
@@ -4,14 +4,45 @@
...
+namespace set_15 {
+
+ov::OutputVector shape(const ov::frontend::onnx::Node& node) {
+    using ov::op::util::is_null;
+
+    const auto data = node.get_ov_inputs().at(0);
+    auto input_shape = std::make_shared<v3::ShapeOf>(data);
+
+    auto start_val = node.get_attribute_value<int64_t>("start", 0);
+    auto end_val = node.get_attribute_value<int64_t>("end", INT64_MAX);
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/shape.cpp:31]: Detail, actually it's worth to use `const` where possible. (Applicable to all of the variables).
```suggestion
    const auto input_shape = std::make_shared<v3::ShapeOf>(data);

    const auto start_val = node.get_attribute_value<int64_t>("start", 0);
    const auto end_val = node.get_attribute_value<int64_t>("end", INT64_MAX);
```

File: src/frontends/onnx/frontend/src/op/shape.cpp [hunk 6]
```diff
@@ -4,14 +4,45 @@
...
+
+    if (start_val == 0 && end_val == INT64_MAX) {
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/shape.cpp:33]: Consider usage of `std::numeric_limits<int64_t>::max()`
instead of `INT64_MAX` macro.

**Review by mitruska — APPROVED** [review]
Now it looks really great, optimized (no redundant Slice if no shape slicing) and backward compatible. Many test cases unblocked 🚀 

(Optional minor comments)

**Comment by gkrivor** [comment]
build_jenkins

**Comment by mlukasze** [comment]
build_jenkins

**Review by gkrivor — APPROVED** [review]

**Pull request merged**